core/src/main/java/org/neo4j/graphalgo/core/utils/LongMinPriorityQueue.java

```diff
@@ -0,0 +1,198 @@
+package org.neo4j.graphalgo.core.utils;
+
+import com.carrotsearch.hppc.IntDoubleScatterMap;
+import com.carrotsearch.hppc.LongDoubleScatterMap;
+import org.apache.lucene.util.ArrayUtil;
+import org.neo4j.collection.primitive.PrimitiveIntIterable;
+import org.neo4j.collection.primitive.PrimitiveIntIterator;
+import org.neo4j.collection.primitive.PrimitiveLongIterable;
+import org.neo4j.collection.primitive.PrimitiveLongIterator;
+
+import java.util.NoSuchElementException;
+
+
+/**
+ * A PriorityQueue specialized for ints that maintains a partial ordering of
+ * its elements such that the smallest value can always be found in constant time.
+ * Put()'s and pop()'s require log(size) time but the remove() cost implemented here is linear.
+ * <p>
+ * <b>NOTE</b>: Iteration order is not specified.
+ *
+ * @author phorn@avantgarde-labs.de
+ */
+public final class LongMinPriorityQueue implements PrimitiveLongIterable {
+    private static final int DEFAULT_CAPACITY = 14;
+    private int size = 0;
+    private long[] heap;
+    private final LongDoubleScatterMap costs;
+
+    public LongMinPriorityQueue() {
+        this(DEFAULT_CAPACITY);
+    }
+
+    public LongMinPriorityQueue(final int initialCapacity) {
+        final int heapSize;
+        if (0 == initialCapacity) {
+            // We allocate 1 extra to avoid if statement in top()
+            heapSize = 2;
+        } else {
+            // NOTE: we add +1 because all access to heap is
+            // 1-based not 0-based.  heap[0] is unused.
+            heapSize = initialCapacity + 1;
+        }
+        this.heap = new long[ArrayUtil.oversize(heapSize, Integer.BYTES)];
+        this.costs = new LongDoubleScatterMap(heapSize);
+    }
+
+    private boolean lessThan(long a, long b) {
+        return costs.get(a) < costs.get(b);
+    }
+
+    /**
+     * Adds an int associated with the given weight to a queue in log(size) time.
+     *
+     * @return the new 'top' element in the queue.
+     */
+    public long add(long element, double cost) {
+        size++;
+        ensureCapacityForInsert();
+        heap[size] = element;
+        costs.addTo(element, cost);
+        upHeap(size);
+        return heap[1];
+    }
+
+    /**
+     * @return the least element of the queue in constant time.
+     */
+    public long top() {
+        // We don't need to check size here: if maxSize is 0,
+        // then heap is length 2 array with both entries null.
+        // If size is 0 then heap[1] is already null.
+        return heap[1];
```

Comment:
but null is not assignable to Long ? so perhaps use an indicator line Long.MIN_VALUE ?